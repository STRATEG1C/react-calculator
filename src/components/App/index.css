* {
    margin: 0;
    padding: 0;
    box-sizing: border-box;
}

html,
body,
#root {
    width: 100%;
    height: 100%;
    font-family: Roboto, sans-serif;
}

#root {
    display: flex;
    justify-content: center;
    align-items: center;
}

.calculator-container {
    width: 450px;
    margin: auto;
    padding: 20px;
    border: 1px solid black;
    border-radius: 30px;
}
