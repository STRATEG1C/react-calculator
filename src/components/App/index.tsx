import React, { useState } from 'react';
import Controls from '../Controls';
import Screen from '../Screen';
import './index.css';
import { removeSpaces } from '../../helpers';

export const App: React.FC = () => {
  const [text, setText] = useState('');

  const calculate = () => {
    try {
      if (text.includes('/0')) {
        throw new Error();
      }
      if (text !== 'Error') {
        setText(eval(text).toString());
      }
    } catch (err) {
      setText('Error');
    }
  }

  const addToText = (value: string) => {
    switch (value) {
      case 'AC': setText(''); break;
      case 'Del': setText((prev) => prev.slice(0, -1)); break;
      case '=': calculate(); break;
      default: setText((prev) => `${prev}${value}`);
    }
  }

  const inputText = (text: string) => {
    setText(removeSpaces(text));
  }

  return (
    <div className="calculator-container">
      <Screen text={text} onInput={inputText} />
      <Controls onClickControl={addToText} />
    </div>
  );
}

export default App;
