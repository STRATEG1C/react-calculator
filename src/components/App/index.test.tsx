import React from 'react';
import { render } from '@testing-library/react';
import App from './index';
import userEvent from '@testing-library/user-event';

describe('Test Application', () => {
  const buttonsToTypeExpression = ['2', '+', '2', '*', '2'];
  const buttonsToTypeWrongExpression = ['2', '/', '0'];
  const EXPECTED_CALCULATED_VALUE = "6";
  const ERROR_VALUE = 'Error';

  it('Should be rendered', () => {
    const { container } = render(<App />);
    expect(container).toMatchSnapshot();
  });

  it('Should input text on button clicks', () => {
    const { getByTestId, getByText } = render(<App />);
    const screen = getByTestId('screen-input') as HTMLInputElement;

    expect(screen.value).toBe('');

    buttonsToTypeExpression.forEach((text) => {
      const button = getByText(text);
      userEvent.click(button);
    });

    expect(screen.value).toBe(buttonsToTypeExpression.join(''));
  });

  it('Should delete text if click on Del button', () => {
    const { getByTestId, getByText } = render(<App />);
    const screen = getByTestId('screen-input') as HTMLInputElement;

    expect(screen.value).toBe('');

    userEvent.click(getByText('1'));
    userEvent.click(getByText('2'));
    userEvent.click(getByText('Del'));

    expect(screen.value).toBe('1');
  });

  it('Should remove all text if click on AC button', () => {
    const { getByTestId, getByText } = render(<App />);
    const screen = getByTestId('screen-input') as HTMLInputElement;

    expect(screen.value).toBe('');

    userEvent.click(getByText('1'));
    userEvent.click(getByText('2'));
    userEvent.click(getByText('AC'));

    expect(screen.value).toBe('');
  });

  it('Should calculate values', () => {
    const { getByTestId, getByText } = render(<App />);
    const screen = getByTestId('screen-input') as HTMLInputElement;
    const makeCalculationButton = getByText('=');

    expect(screen.value).toBe('');

    buttonsToTypeExpression.forEach((text) => {
      const button = getByText(text);
      userEvent.click(button);
    });

    userEvent.click(makeCalculationButton);

    expect(screen.value).toBe(EXPECTED_CALCULATED_VALUE);
  });

  it('Should give an error while division on zero', () => {
    const { getByTestId, getByText } = render(<App />);
    const screen = getByTestId('screen-input') as HTMLInputElement;
    const makeCalculationButton = getByText('=');

    expect(screen.value).toBe('');

    buttonsToTypeWrongExpression.forEach((text) => {
      const button = getByText(text);
      userEvent.click(button);
    });

    userEvent.click(makeCalculationButton);

    expect(screen.value).toBe(ERROR_VALUE);
  });
});
