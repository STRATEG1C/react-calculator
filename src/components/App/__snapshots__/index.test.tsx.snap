// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Application Should be rendered 1`] = `
<div>
  <div
    class="calculator-container"
  >
    <div
      class="screen"
    >
      <input
        class="screen-input"
        data-testid="screen-input"
        value=""
      />
    </div>
    <div
      class="controls"
    >
      <div
        class="row"
      >
        <div
          class="button operator"
        >
          AC
        </div>
        <div
          class="button operator"
        >
          Del
        </div>
        <div
          class="button operator"
        >
          *
        </div>
        <div
          class="button operator"
        >
          /
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="button "
        >
          7
        </div>
        <div
          class="button "
        >
          8
        </div>
        <div
          class="button "
        >
          9
        </div>
        <div
          class="button operator"
        >
          +
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="button "
        >
          4
        </div>
        <div
          class="button "
        >
          5
        </div>
        <div
          class="button "
        >
          6
        </div>
        <div
          class="button operator"
        >
          -
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="button "
        >
          1
        </div>
        <div
          class="button "
        >
          2
        </div>
        <div
          class="button "
        >
          3
        </div>
        <div
          class="button operator-equal"
        >
          =
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="button invisible"
        />
        <div
          class="button "
        >
          0
        </div>
        <div
          class="button invisible"
        />
      </div>
    </div>
  </div>
</div>
`;
